widget/hexwidget/hexwidget.go

```diff
@@ -0,0 +1,214 @@
+}
+
+func (h *HexWidget) getSegmentColor(segno int) color.RGBA {
+	if (h.segments & (1 << segno)) == 0 {
+		return hexOnColor
+	}
+
+	return hexOffColor
+}
+
+// CreateRenderer implements fyne.Widget
+func (h *HexWidget) CreateRenderer() fyne.WidgetRenderer {
+	seg0 := canvas.NewLine(hexOffColor)
+	seg1 := canvas.NewLine(hexOffColor)
+	seg2 := canvas.NewLine(hexOffColor)
+	seg3 := canvas.NewLine(hexOffColor)
+	seg4 := canvas.NewLine(hexOffColor)
+	seg5 := canvas.NewLine(hexOffColor)
+	seg6 := canvas.NewLine(hexOffColor)
+
+	seg0.StrokeWidth = float32(hexSegmentWidth / 2)
+	seg1.StrokeWidth = float32(hexSegmentWidth / 2)
+	seg2.StrokeWidth = float32(hexSegmentWidth / 2)
+	seg3.StrokeWidth = float32(hexSegmentWidth / 2)
+	seg4.StrokeWidth = float32(hexSegmentWidth / 2)
+	seg5.StrokeWidth = float32(hexSegmentWidth / 2)
+	seg6.StrokeWidth = float32(hexSegmentWidth / 2)
+
+	pos := image.Pt(0, 0)
+
+	// pt0Center := image.Pt(pos.X+int(hexWidth/2.0+hexOffset), pos.Y+int((hexHeight-2.0*hexSegmentVLength)/2))
+	pt0Center := image.Pt(pos.X+int(hexWidth/2.0+hexOffset), pos.Y)
+	pt05 := image.Pt(int(float32(pt0Center.X)-(hexSegmentHLength/2)), pt0Center.Y)
+	pt01 := image.Pt(int(float32(pt0Center.X)+(hexSegmentHLength/2)), pt0Center.Y)
+
+	pt6Center := image.Pt(pos.X+int(hexWidth/2.0), int(float32(pt0Center.Y)+hexSegmentVLength))
+	pt65 := image.Pt(int(float32(pt6Center.X)-(hexSegmentHLength/2)), pt6Center.Y)
+	pt61 := image.Pt(int(float32(pt6Center.X)+(hexSegmentHLength/2)), pt6Center.Y)
+
+	pt3Center := image.Pt(pos.X+int(hexWidth/2.0-hexOffset), int(float32(pt0Center.Y)+2*hexSegmentVLength))
+	pt34 := image.Pt(int(float32(pt3Center.X)-(hexSegmentHLength/2)), pt3Center.Y)
+	pt32 := image.Pt(int(float32(pt3Center.X)+(hexSegmentHLength/2)), pt3Center.Y)
+
+	// segment 0
+	setLineEndpoints(seg0, pt05, pt01)
+
+	// segment 1
+	setLineEndpoints(seg1, pt01, pt61)
+
+	// segment 2
+	setLineEndpoints(seg2, pt61, pt32)
+
+	// segment 3
+	setLineEndpoints(seg3, pt32, pt34)
+
+	// segment 4
+	setLineEndpoints(seg4, pt34, pt65)
+
+	// segment 5
+	setLineEndpoints(seg5, pt65, pt05)
+
+	// segment 6
+	setLineEndpoints(seg6, pt65, pt61)
+
+	return &hexRenderer{
+		hex:            h,
+		segmentObjects: []fyne.CanvasObject{seg0, seg1, seg2, seg3, seg4, seg5, seg6},
+	}
+}
+
+// NewHexWidget instantiates a new widget instance, with all of the segments
+// disabled.
+func NewHexWidget() *HexWidget {
+	h := &HexWidget{
+		segments: 0xff,
+	}
+
+	h.ExtendBaseWidget(h)
+	return h
+}
+
+// UpdateSegments changes the state of the segments and causes the widget to
+// refresh so the changes are visible to the user.
+func (h *HexWidget) UpdateSegments(segments uint8) {
+	h.segments = segments
+	h.Refresh()
+}
+
+// Set updates the hex widget to show a specific number between 0 and 15, which
+// will be rendered in hexadecimal in 0...f. If the number is greater than 15,
+// it will be modulo-ed by 16.
+func (h *HexWidget) Set(val uint) {
```

Comment:
Random thought, would it be nicer to use `rune` OR would you consider consts for the `a`, rather than developer passing `10`?

Reply:
I don't think so. It can't really handle arbitrary runes. Also, Go allows for hexadecimal values. You can give run `h.Set(0xa)` and it will do the right thing. This way makes it easy to slice apart a bigger number with masks and set each corresponding segment individually. Using rune would make it require string processing, and consts would mean the caller would have to look stuff up.